lightning/restore/restore.go

```diff
@@ -304,48 +305,210 @@ outside:
+					if err != nil {
+						worker.throw(err)
+					}
+					for _, sql := range stmts {
+						restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+							tblName:  viewMeta.Name,
+							stmtType: schemaCreateView,
+							sql:      sql,
+						})
+					}
+				}
 			}
+			worker.wg.Add(1)
+			worker.jobCh <- restoreSchemaJob
+			// we don't support restore views concurrency, cauz it maybe will raise a error
+			worker.wg.Wait()
 		}
+		worker.quit()
+	}
+}
 
-		// restore views. Since views can cross database we must restore views after all table schemas are restored.
-		for _, dbMeta := range rc.dbMetas {
-			if len(dbMeta.Views) > 0 {
-				task := log.With(zap.String("db", dbMeta.Name)).Begin(zap.InfoLevel, "restore view schema")
-				viewsSchema := make(map[string]string)
-				for _, viewMeta := range dbMeta.Views {
-					viewsSchema[viewMeta.Name] = viewMeta.GetSchema(ctx, rc.store)
+func (worker *restoreSchemaWorker) doJob() {
+	for {
+		select {
+		case <-worker.ctx.Done():
+			return
+		case job := <-worker.jobCh:
+			if job == nil {
+				return
+			}
+			session := worker.getSession()
+			var logger log.Logger
+			for _, stmt := range job.stmts {
+				if stmt.stmtType == schemaCreateDatabase {
+					logger = log.With(zap.String("db", job.dbName))
+				} else if stmt.stmtType == schemaCreateTable {
+					logger = log.With(zap.String("table", common.UniqueTable(job.dbName, stmt.tblName)))
+				} else if stmt.stmtType == schemaCreateView {
+					logger = log.With(zap.String("table", common.UniqueTable(job.dbName, stmt.tblName)))
 				}
-				err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, viewsSchema)
-
+				task := logger.Begin(zap.DebugLevel, fmt.Sprintf("execute SQL: %s", stmt.sql))
+				_, err := session.Execute(worker.ctx, stmt.sql)
 				task.End(zap.ErrorLevel, err)
 				if err != nil {
-					return errors.Annotatef(err, "restore view schema %s failed", dbMeta.Name)
+					switch stmt.stmtType {
+					case schemaCreateDatabase:
+						err = errors.Annotatef(err, "restore database schema %s failed", job.dbName)
+					case schemaCreateTable:
+						err = errors.Annotatef(err, "restore table schema %s failed", stmt.tblName)
+					case schemaCreateView:
+						err = errors.Annotatef(err, "restore view schema %s failed", stmt.tblName)
+					}
+					worker.recycleSession(session)
+					worker.wg.Done()
+					worker.throw(err)
 				}
 			}
+			worker.recycleSession(session)
+			worker.wg.Done()
+		}
+	}
+}
+
+func (worker *restoreSchemaWorker) wait() error {
+	defer func() {
+		worker.quit()
+		close(worker.sessions)
+		for session := range worker.sessions {
+			session.Close()
+		}
+	}()
+	select {
+	case err := <-worker.errCh:
+		return err
+	case <-worker.ctx.Done():
+		return nil
+	}
+}
+
+func (worker *restoreSchemaWorker) throw(err error) {
+	worker.errCh <- err
+}
+
+func (worker *restoreSchemaWorker) getSession() checkpoints.Session {
+	if len(worker.sessions) > 0 {
+		return <-worker.sessions
```

Comment:
please add a comment for this function that it should not be called concurrently for same worker. And I think it's no need to use a channel to lazy initiate session

Reply:
There has two  main reasons:
1.  I use buffered channel is that I cannot use idle connections properly
2. I also need a session pool with the same size of concurrency. 

If don't use buffered channel, purpose 1 will be a little hard to implement.

Reply:
don't know what you mean. idle connection are assigned to a member of worker and used as usual